overlord/snapshotstate/snapshotmgr.go

```diff
@@ -0,0 +1,255 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapshotstate
+
+import (
+	"encoding/json"
+	"os"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/overlord/snapshotstate/backend"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+// SnapshotManager takes snapshots of active snaps
+type SnapshotManager struct {
+	runner *state.TaskRunner
+}
+
+// Manager returns a new SnapshotManager
+func Manager(st *state.State) *SnapshotManager {
+	runner := state.NewTaskRunner(st)
+	runner.AddHandler("save-snapshot", doSave, doForget)
+	runner.AddHandler("forget-snapshot", doForget, nil)
+	runner.AddHandler("check-snapshot", doCheck, nil)
+	runner.AddHandler("restore-snapshot", doRestore, undoRestore)
+	runner.AddCleanup("restore-snapshot", cleanupRestore)
+	return &SnapshotManager{runner: runner}
+}
+
+// KnownTaskKinds is part of the overlord.StateManager interface.
+func (m *SnapshotManager) KnownTaskKinds() []string {
+	return m.runner.KnownTaskKinds()
+}
+
+// Ensure is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Ensure() error {
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Stop() {
+	m.runner.Stop()
+}
+
+type snapshotState struct {
+	SetID    uint64   `json:"set-id"`
+	Snap     string   `json:"snap"`
+	Users    []string `json:"users"`
+	Filename string   `json:"filename"`
+}
+
+func filename(setID uint64, si *snap.Info) string {
+	skel := &client.Snapshot{
+		SetID:    setID,
+		Snap:     si.Name(),
+		Revision: si.Revision,
+		Version:  si.Version,
+	}
+	return backend.Filename(skel)
+}
+
+func prepareSave(task *state.Task) (*snapshotState, *snap.Info, map[string]interface{}, error) {
```

Comment:
Consider using named return values, the same goes for `prepareRestore`

Reply:
this bit was pushed to this pr in error. I'm fixing it with a forced push. Please don't hate me!